### PR TITLE
Add step to check secure protocols supported by server

### LINKED ISSUE #681: Add step to verify the TLS protocol versions supported by server
**Is your feature request related to a problem? Please describe.**
Server could support TLSv1.1 and TLSv1.0 protocols which is not recommended using for now, therefore we should verify that among all TLS protocols which are currently enabled for use on connection is not those version. 

**Describe the solution you'd like**
Then the number of the TLS protocol version with the name 'protocolName' is = 0
Then the number of the TLS protocol version with the name 'protocolName' is >= 1

As an option for implementing this feature maybe the method getEnabledProtocols() from  SSLSocket class in javax.net.ssl package can help you. (SSLSocketFactory() -> createSocket(String host, int port))

### PR DESCRIPTION
**Pull request by uarlouski**
Closes #681

### CODE REVIEW

File: vividus-http-client/src/main/java/org/vividus/http/client/SslContextFactory.java [hunk 1]
```diff
@@ -93,4 +96,10 @@ public void setPrivateKeyPassword(String privateKeyPassword)
     {
         this.privateKeyPassword = privateKeyPassword;
     }
+
+    @FunctionalInterface
+    private interface GeneralSecurityExceptionSupplier<T>
```
Comment (valfirst) [inline vividus-http-client/src/main/java/org/vividus/http/client/SslContextFactory.java:101]: org.vividus.util.function.CheckedSupplier

File: vividus-plugin-rest-api/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker [hunk 2]
```diff
@@ -0,0 +1 @@
+mock-maker-inline
```
Comment (valfirst) [inline vividus-plugin-rest-api/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker:1]: why do you need this change?

File: vividus-tests/src/main/resources/story/integration/SslSteps.story [hunk 3]
```diff
@@ -0,0 +1,9 @@
+Description: Integration tests for SslSteps functionality
+
+Meta:
+    @epic vividus-plugin-rest-api
+
+Scenario: Verify step: "Then server `$hostname` supports secure protocols that $rule `$protocols`"
+Meta:
+    @issueId 681
+Then server `example.com` supports secure protocols that CONTAIN `TLSv1.3,TLSv1.2`
```
Comment (valfirst) [inline vividus-tests/src/main/resources/story/integration/SslSteps.story:9]: can we use `https://vividus-test-site.herokuapp.com/` ?

Comment (valfirst) [inline vividus-tests/src/main/resources/story/integration/SslSteps.story:9]: `CONTAIN` -> `contain`

File: vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/ssl/SslSteps.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+     */
+    @Then("server `$hostname` supports secure protocols that $rule `$protocols`")
+    public void checkSupportedSecureProtocols(String hostname, ListComparisonRule rule, List<String> protocols)
+            throws GeneralSecurityException, IOException
+    {
+        SSLSocketFactory socketFactory = sslContextFactory.getDefaultSslContext().getSocketFactory();
+        try (SSLSocket socket = (SSLSocket) socketFactory.createSocket(hostname, SSL_PORT))
+        {
+            List<String> enabledProtocols = List.of(socket.getEnabledProtocols());
+            softAssert.assertThat("Enabled secure protocols", enabledProtocols, rule.getComparisonRule(protocols));
```
Comment (valfirst) [inline vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/ssl/SslSteps.java:62]: what about overloading method `getComparisonRule` to accept also array as input parameter?

File: vividus-bdd-engine/src/main/java/org/vividus/bdd/steps/ListComparisonRule.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+package org.vividus.bdd.steps;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.hamcrest.Matcher;
+import org.hamcrest.Matchers;
+
+public enum ListComparisonRule
```
Comment (valfirst) [inline vividus-bdd-engine/src/main/java/org/vividus/bdd/steps/ListComparisonRule.java:25]: would you mind adding the documentation for this rule?

File: vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/ssl/SslSteps.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+     *
+     * @param hostname server host e.g. example.com
+     * @param rule one of the CONTAIN, ARE_EQUAL_TO, ARE_EQUAL_TO_ORDERED_LIST rules
+     * @param protocols comma separated list of protocols
+     * @throws GeneralSecurityException If a security exception occurred
+     * @throws IOException If an input or output exception occurred
+     */
+    @Then("server `$hostname` supports secure protocols that $rule `$protocols`")
+    public void checkSupportedSecureProtocols(String hostname, ListComparisonRule rule, List<String> protocols)
+            throws GeneralSecurityException, IOException
```
Comment (valfirst) [inline vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/ssl/SslSteps.java:56]: `GeneralSecurityException` is not thrown

File: vividus-bdd-engine/src/main/java/org/vividus/bdd/steps/ListComparisonRule.java [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    CONTAIN
+    {
+        @SuppressWarnings({ "unchecked", "rawtypes" })
+        @Override
+        public <T> Matcher<Iterable<T>> getComparisonRule(List<T> values)
+        {
+            return values.stream()
+                         .map(Matchers::hasItem)
+                         .collect(Collectors.collectingAndThen(Collectors.toList(),
+                             m -> Matchers.allOf((List) m)));
```
Comment (valfirst) [inline vividus-bdd-engine/src/main/java/org/vividus/bdd/steps/ListComparisonRule.java:36]: `return Matchers.hasItems((T[]) values.toArray(new Object[0]));`

Comment (valfirst) [reply vividus-bdd-engine/src/main/java/org/vividus/bdd/steps/ListComparisonRule.java:36]: the same for another items

File: vividus-plugin-rest-api/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker (hunk 2 above)
Comment (uarlouski) [reply vividus-plugin-rest-api/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker:1]: https://github.com/mockito/mockito/wiki/What%27s-new-in-Mockito-2#mock-the-unmockable-opt-in-mocking-of-final-classesmethods
SSLContext.getSocketFactory() is final

File: vividus-bdd-engine/src/main/java/org/vividus/bdd/steps/CollectionComparisonRule.java [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+                         .collect(Collectors.collectingAndThen(Collectors.toList(),
+                             m -> Matchers.allOf((Iterable) m)));
+        }
+    },
+    ARE_EQUAL_TO
+    {
+        @Override
+        public <T> Matcher<Iterable<T>> getComparisonRule(List<T> items)
+        {
+            return Matchers.allOf(Matchers.containsInAnyOrder(asArray(items)));
```
[the remark this answers is not in the document]
Comment (uarlouski) [reply vividus-bdd-engine/src/main/java/org/vividus/bdd/steps/CollectionComparisonRule.java:51]: containsInAnyOrder returns Iterable<? extends T> while hasItems returns Iterable<T>, so I use allOf to align types

[the remark this answers is not in the document]
Comment (valfirst) [reply vividus-bdd-engine/src/main/java/org/vividus/bdd/steps/CollectionComparisonRule.java:51]: `public <T> Matcher<Iterable<? extends T>> getComparisonRule(List<T> items)` ?

[the remark this answers is not in the document]
Comment (uarlouski) [reply vividus-bdd-engine/src/main/java/org/vividus/bdd/steps/CollectionComparisonRule.java:51]: Forgot to add type param for hasItems, that's hasItems returns Iterable<T>

File: vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/ssl/SslSteps.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+    /**
+     * Checks that a server defined by the <b>hostname</b> supports secure protocols defined by the
+     * <b>protocols</b> parameter.
+     *
+     * @param hostname server host e.g. example.com
+     * @param rule one of the CONTAIN, ARE_EQUAL_TO, ARE_EQUAL_TO_ORDERED_COLLECTION rules
+     * @param protocols comma separated list of protocols
+     * @throws IOException If an input or output exception occurred
+     */
+    @Then("server `$hostname` supports secure protocols that $rule `$protocols`")
```
Comment (valfirst) [inline vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/ssl/SslSteps.java:51]: would you mind adding the documentation for this step?

File: docs/modules/plugins/pages/plugin-rest-api.adoc [hunk 10]
```diff
@@ -1 +1,49 @@
...
+.build.gradle
+[source,gradle]
+----
+implementation(group: 'org.vividus', name: 'vividus-plugin-rest-api', version: '0.2.3')
+----
+
+
+== Properties
+
+TO-DO
```
Comment (ikalinin1) [inline docs/modules/plugins/pages/plugin-rest-api.adoc:19]: [WARNING]
To be implemented 

?

File: vividus-bdd-engine/src/test/java/org/vividus/bdd/steps/CollectionComparisonRuleTests.java [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+import java.util.List;
+import java.util.stream.Stream;
+
+import org.junit.jupiter.api.DynamicTest;
+import org.junit.jupiter.api.TestFactory;
+
+class CollectionComparisonRuleTests
+{
+    @TestFactory
+    Stream<DynamicTest> listRules()
```
Comment (ikalinin1) [inline vividus-bdd-engine/src/test/java/org/vividus/bdd/steps/CollectionComparisonRuleTests.java:35]: From my point of view, it is easier to read using a data source

CONTAIN, List.of(1, 0), List.of(1, 1, 1), false
where
rule, expected, actual, comparisonResult

File: vividus-tests/src/main/resources/story/integration/SslSteps.story [hunk 12]
```diff
@@ -0,0 +1,9 @@
+Description: Integration tests for SslSteps functionality
+
+Meta:
+    @epic vividus-plugin-rest-api
+
+Scenario: Verify step: "Then server `$hostname` supports secure protocols that $rule `$protocols`"
+Meta:
+    @issueId 681
```
Comment (ikalinin1) [inline vividus-tests/src/main/resources/story/integration/SslSteps.story:8]: @requirementId ?

File: vividus-bdd-engine/src/test/java/org/vividus/bdd/steps/CollectionComparisonRuleTests.java (hunk 11 above)
Comment (uarlouski) [reply vividus-bdd-engine/src/test/java/org/vividus/bdd/steps/CollectionComparisonRuleTests.java:35]: depends on :)

**Pull request merged**